src/storage/kv_store.rs

```diff
@@ -335,50 +342,64 @@ pub struct StorageHandler {
 }
 
 impl StorageHandler {
-    pub fn open(path: PathBuf, read_only: bool, memory_budget: u64) -> Self {
+    pub fn open(path: PathBuf, read_only: bool, config_memory_budget: Option<u64>) -> Self {
         info!("opening db...");
         let mut db_opts = Options::default();
         db_opts.create_missing_column_families(true);
         db_opts.create_if_missing(true);
 
-        info!(
-            "using rocksdb memory budget: {:.2} GB ({} bytes)",
-            memory_budget as f64 / 1024.0 / 1024.0 / 1024.0,
-            memory_budget
-        );
+        // Enable RocksDB statistics for monitoring
+        db_opts.enable_statistics();
+        db_opts.set_report_bg_io_stats(true);
 
-        // Create a shared cache for both block cache and write buffer manager
-        let write_buffer_budget = (memory_budget as f64 * 0.3) as usize;
+        let memory_budget = config_memory_budget.unwrap_or(default_rocksdb_memory_budget());
 
-        let cache = Cache::new_lru_cache(memory_budget.try_into().unwrap());
-        let write_buffer_manager = WriteBufferManager::new_write_buffer_manager_with_cache(
-            write_buffer_budget,
-            false,
-            cache.clone(),
-        );
+        let memtable_budget = (memory_budget as f64 * 0.25) as u64;
+        let per_memtable_cap = 512 * 1024 * 1024;
+        let write_buffer_size = std::cmp::min(memtable_budget / 2, per_memtable_cap);
+        let memtable_budget = write_buffer_size * 2;
 
-        // Set the write buffer manager to control total memtable memory
-        db_opts.set_write_buffer_manager(&write_buffer_manager);
+        // remaining budget on block cache
+        let block_cache_budget = memory_budget - memtable_budget;
```

Comment:
Lets try to explain these budgets and thresholds a bit, perhaps in a dedicated docs in `docs/`, and why we're splitting them up the way we are. Doesn't have to be very profound or extensive but would be good to clarify and justify.